### PR TITLE
Revise the exemplar default reservoirs

### PR DESCRIPTION
**Pull request by MrAlias**
- Make it normative that an SDK is recommended to include both types of default reservoirs.
- Make the description of both default reservoirs their own sub-sections.
- Correct the "SimpleExemplarReservoir" section to be titled "SimpleFixedSizeExemplarReservoir".
- Loosen the SimpleFixedSizeExemplarReservoir to allow any uniformly-weighted sampling algorithm instead of just the naive one that is not optimal.
  - The current algorithm has an asymptotic running time of `O(n)`. There are other sampling algorithms that produce equivalent results and have more optimal asymptotic running times. Therefore, do not restrict implementations to only implementing the inefficient algorithm.

### CODE REVIEW

File: specification/metrics/sdk.md [hunk 1]
```diff
@@ -994,10 +997,15 @@ algorithm](https://en.wikipedia.org/wiki/Reservoir_sampling)
   end
   ```
 
-Additionally, the `num_measurements_seen` count SHOULD be reset at every
-collection cycle.
+The reservoir SHOULD be reset for every collection cycle. So, following the
+above example, the `num_measurements_seen` count would be reset every time the
+reservoir is collected.
+
+This Exemplar reservoir MAY take a configuration parameter for the size of the
```
Comment (cijothomas) [inline specification/metrics/sdk.md:1004]: would we also be specifying a default value for the reservoir pool size?

**Review by cijothomas — APPROVED** [review]

**Review by reyang — APPROVED** [review]

File: specification/metrics/sdk.md (hunk 1 above)
Comment (MrAlias) [reply specification/metrics/sdk.md:1004]: Yeah, good question. This is an open question being tracked in https://github.com/open-telemetry/opentelemetry-specification/issues/2421.

I don't plan to answer this question in this PR, but am happy to open another PR if we can reach consensus in that issue as to what the default should be.

Comment (MrAlias) [reply specification/metrics/sdk.md:1004]: Going to resolve the comment given I don't plan to address here. Please unresolve if you think this is an error.

File: specification/metrics/sdk.md [hunk 2]
```diff
@@ -972,20 +974,21 @@ The `ExemplarReservoir` SHOULD avoid allocations when sampling exemplars.
...
 `SimpleFixedSizeExemplarReservoir`.
 
-_SimpleExemplarReservoir_
-This Exemplar reservoir MAY take a configuration parameter for the size of the
-reservoir pool.  The reservoir will accept measurements using an equivalent of
-the [naive reservoir sampling
-algorithm](https://en.wikipedia.org/wiki/Reservoir_sampling)
+#### SimpleFixedSizeExemplarReservoir
+
+This reservoir MUST use an uniformly-weighted sampling algorithm to determine
```
Comment (jsuereth) [inline specification/metrics/sdk.md:988]: The original definition wasn't meant to lock down algorithm, for what it's worth.  

Two points:

1. I'm not sure how this lead to O(N) performance, so I'd actually like to understand that better.
2. Can you improve the language to denote the "weight" used is based on the number of samples the reservoir has seen so far?

File: specification/metrics/sdk.md [hunk 3]
```diff
@@ -994,10 +997,15 @@ algorithm](https://en.wikipedia.org/wiki/Reservoir_sampling)
   end
   ```
 
-Additionally, the `num_measurements_seen` count SHOULD be reset at every
-collection cycle.
+The reservoir SHOULD be reset for every collection cycle. So, following the
```
Comment (jsuereth) [inline specification/metrics/sdk.md:1000]: This is a good clarification.  Maybe expand the clarification to be something like:

```
Any stateful portion of sampling computation SHOULD also be reset.  For the above example, that would mean that `num_measurements_seen` count is reset every time the reservoir is collected.
```

**Review by jsuereth — APPROVED** [review]
Overall like these improvements, just two minor nits.

File: specification/metrics/sdk.md (hunk 2 above)
Comment (MrAlias) [reply specification/metrics/sdk.md:988]: > 1. I'm not sure how this lead to O(N) performance, so I'd actually like to understand that better.

As mentioned in the linked wiki ...

> [t]his algorithm needs to generate a random number for each item of the input, including the items that are discarded. The algorithm's asymptotic running time is thus O(n).

Comment (MrAlias) [reply specification/metrics/sdk.md:988]: > 2\. Can you improve the language to denote the "weight" used is based on the number of samples the reservoir has seen so far?

Sure, will do :+1:

Comment (jsuereth) [reply specification/metrics/sdk.md]: Ah, I see, you meant O(N) over the whole input, whereas I was thinking the complexity per-measurement.

**Pull request merged**